nexus/src/internal_api/http_entrypoints.rs

```diff
@@ -540,3 +544,79 @@ async fn bgtask_view(
     };
     apictx.internal_latencies.instrument_dropshot_handler(&rqctx, handler).await
 }
+
+// NAT RPW internal APIs
+
+/// Path parameters for NAT ChangeSet
+#[derive(Deserialize, JsonSchema)]
+struct RpwNatPathParam {
+    /// which change number to start generating
+    /// the change set from
+    from_gen: u32,
+}
+
+/// Query parameters for NAT ChangeSet
+#[derive(Deserialize, JsonSchema)]
+struct RpwNatQueryParam {
+    limit: u32,
+}
+
+/// Fetch NAT ChangeSet
+///
+/// Caller provides their generation as `from_gen`, along with a query
+/// parameter for the page size (`limit`). Endpoint will return changes
+/// that have occured since the caller's generation number up to the latest
+/// change or until the `limit` is reached. If there are no changes, an
+/// empty vec is returned.
+#[endpoint {
+   method = GET,
+    path = "/rpw/nat/ipv4/changeset/{from_gen}"
+}]
+async fn ipv4_nat_changeset(
+    rqctx: RequestContext<Arc<ServerContext>>,
+    path_params: Path<RpwNatPathParam>,
+    query_params: Query<RpwNatQueryParam>,
+) -> Result<HttpResponseOk<Vec<Ipv4NatEntryView>>, HttpError> {
+    let apictx = rqctx.context();
+    let handler = async {
+        let opctx = crate::context::op_context_for_internal_api(&rqctx).await;
+        let nexus = &apictx.nexus;
+        let path = path_params.into_inner();
+        let query = query_params.into_inner();
+        let mut changeset = nexus
+            .datastore()
+            .ipv4_nat_changeset(&opctx, path.from_gen, query.limit)
+            .await?;
+        changeset.sort_by_key(|e| e.gen);
+        Ok(HttpResponseOk(changeset))
+    };
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+// TODO: Remove (maybe)
+// This endpoint may be more trouble than it's worth. The reason this endpoint
+// may be a headache is because the underlying SEQUENCE in the DB will advance,
+// even on failed transactions. This is not a big deal, but it will lead to
+// callers trying to catch up to entries that don't exist if they check this
+// endpoint first. As an alternative, we could just page through the
+// changesets until we get nothing back, then revert to periodic polling
+// of the changeset endpoint.
```

Comment:
Where'd we land on this?

Reply:
I haven't seen a place where it has merit, so I'll remove it.